### PR TITLE
feat(api,client): Add Webhooks with Optional Token

### LINKED ISSUE #52542: Webhook to Emit Events When a Camper Completes Challenges
### Is your feature request related to a problem? Please describe.

Some platforms, such as Github, include an optional webhooks architecture so that users can emit events that other platforms can consume. This can be used to trigger notifications, track students assigned to use FCC as curriculum and any number of other uses. 

Enabling a webhook would allow educational programs to utilize FCC as some of their curriculum while also tracking every student's engagement and accomplishment. I am an educator and this feature request comes from my own desire to better understand where my students are getting stuck and whether they are doing the work at all. 

Here is a screenshot of how and where to add a webhook for a repo on Github as an example:

Screen Shot 2023-12-13 at 3 53 14 PM

### Describe the solution you'd like

I'm happy to build this feature and I've attached some screen shots of what I have started on. 

Under the user settings, there will be a section for a user to add a URL as a destination for the webhook. 

Screen Shot 2023-12-13 at 3 49 26 PM

As an example, a Camper with a webhook enabled, working on the Responsive Web Dev track, starting with the Building a Cat Photo App would have an even emitted every time they completed a step. That would look like the following sequence

1.) Write code to satisfy the challenge
2.) Click 'Check Your Code'
3.) Click 'Submit and go to next challenge'
4.) After progress is registered on the BE, a webhook is emitted that would look like this: 
```
  {
    username: 'certifieduser',
    completedChallenge: {
      id: '5dc174fcf86c76b9248c6eb2',
      files: undefined,
      completedDate: 1702499864614
    },
    points: 1
  }
```

This allows the receiver to mark the identified challenge as complete for the user, tabulate their points, etc. 

### Describe alternatives you've considered

One alternative would be to develop an API to surface more Camper data however this doesn't allow for active pushed updates when events are completed and would require as much or more overhead. Finally, an external API is at more risk of abuse.

### Additional context

Another platform that offers a feature like this is www.Codewars.com. You can see it in the [user settings](https://www.codewars.com/users/edit) or you can see the documentation for it [here](https://dev.codewars.com/#webhooks). 

Screen Shot 2023-12-13 at 3 07 08 PM

For me, I know this would be a very useful feature and I think that many other organizations will find creative ways to leverage this as well. 

There are some more specs that need to be decided for the feature but I think this covers the basic idea. I'm happy to answer any questions and I'd love to hear any one else's ideas about this. 

Thanks!

### PR DESCRIPTION
**Pull request by werner33**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #52542

This adds the following functionality: 

1) A form on the settings page where a user can add a URL (webhook) and an optional token.

2) When the token is added and when a challenge is completed, a small object is emitted to the URL provided. 

3) The webhook can be updated or deleted at any time on the settings page.

### CODE REVIEW

**Comment by ojeytonwilliams** [comment]
Hi @werner33 thanks a ton for putting this together, it really helps to see the implementation. 

We've started discussing this internally, but it will take us a little time to figure out if this is something we can support. Typically we meet weekly, so, all being well, we should be able to get back to you next week. :crossed_fingers:

**Comment by werner33** [comment]
Thanks so much @ojeytonwilliams. I know more changes or refinements may be needed and I am happy to do them. Please let me know if you would like anything else.

File: api-server/src/server/boot/settings.js [hunk 1]
```diff
@@ -334,6 +345,24 @@ function updateMyQuincyEmail(...args) {
   )(...args);
 }
 
+function updateMyWebhook(...args) {
+  const buildUpdate = body => _.pick(body, ['webhook', 'webhookSecret']);
+  const validate = ({ webhook }) => isURL(webhook, { require_protocol: true });
```
Comment (ShaunSHamilton) [inline api-server/src/server/boot/settings.js:350]: We should require `https` here.

File: client/src/components/settings/webhook.tsx [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    const { newWebhook } = webhookForm;
+
+    if (!newWebhook) {
+      return {
+        state: null,
+        message: ''
+      };
+    }
+
+    if (!isURL(newWebhook)) {
```
Comment (ShaunSHamilton) [inline client/src/components/settings/webhook.tsx:88]: Validation should match server.

File: api-server/src/server/utils/middleware.js [hunk 3]
```diff
@@ -106,3 +106,36 @@ export const createValidatorErrorHandler =
...
+  };
+  const {
+    user: { username },
+    body: { webhook, webhookSecret }
+  } = req;
+
+  try {
+    await triggerWebhook(webhook, webhookSecret, username);
+    next();
+  } catch (error) {
```
Comment (ShaunSHamilton) [inline api-server/src/server/utils/middleware.js:123]: This error is forgotten about. I would send it to the client, and make the client error more specific that the sent resource did not successfully respond.

File: api-server/src/server/utils/middleware.js [hunk 4]
```diff
@@ -106,3 +106,36 @@ export const createValidatorErrorHandler =
...
+    type: 'danger',
+    message: 'flash.wrong-updating'
+  };
+  const {
+    user: { username },
+    body: { webhook, webhookSecret }
+  } = req;
+
+  try {
+    await triggerWebhook(webhook, webhookSecret, username);
```
Comment (ShaunSHamilton) [inline api-server/src/server/utils/middleware.js:121]: URL validation should occur before making first contact.

File: api-server/src/server/boot/challenge.js [hunk 5]
```diff
@@ -406,6 +405,19 @@ export async function modernChallengeCompleted(req, res, next) {
...
+      // emit webhook here
+      fetch(user.webhook, {
+        method: 'POST',
+        body: JSON.stringify({
+          username: user.username,
+          completedChallenge,
+          points
+        }),
+        headers: { 'Content-Type': 'application/json' }
+      });
```
Comment (ShaunSHamilton) [inline api-server/src/server/boot/challenge.js:418]: I cannot think of something more useful than just throwing away any error.

```suggestion
      }).catch(() => {});
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
**Other Changes**:

The client should mention this feature being in "Beta"
  - We want it clear this feature is subject to major change

---

_Thank you for your patience on this_

**Comment by ShaunSHamilton** [comment]
@werner33 We realise this feature has taken longer to get reviewed than desirable. If you do not mind, I can push the changes necessary to resolve my previous review?

**Comment by werner33** [comment]
Please feel free to make any changes. I don’t mind. I made some changes
that you might see there. Thanks so much!

**Comment by ShaunSHamilton** [comment]
@werner33 I am making changes to this PR, and should push them up tomorrow.

For the initial launch, we do want to restrict the URLs allowed in the webhooks. Would you mind sharing the URL **you** are interested in receiving on?

**Comment by raisedadead** [comment]
Hey @werner33 do yo mind replying over on GitHub's UI - it seems the email reply has a bad format. Cheers.

**Comment by werner33** [comment]
@ShaunSHamilton Thank you so much for the work you are doing here. I really appreciate it. 

@raisedadead I saw your message today. I'm sorry I got very sick on the 15th and have only today been back on my computer at all. I think the info is resolved but please do let me know if anything else is needed. 

Thank you both very much!

**Comment by raisedadead** [comment]
Hey @werner33 - No worries – we work async. I just wish you a speedy recovery. If we have any further questions we will let you know. Cheers.

**Comment by raisedadead** [comment]
Hey @werner33 

Thanks a lot for your patience with this. We had a long meeting about this feature. We have unfortunately come to the conclusion that it will add significant challenges in moving this over to the new API.

In an ideal world we want to offload sending webhook events to a "micro-service". That's something I think we do not want to build into the current API.

I am happy to meet you to explain this in detail if you are interested in the technicals.

Thanks again, I think we like the idea of webhooks and we will build this in the new API with the GitHub-style experience you and I talked about.

It should be more powerful and secure that way.

Cheers!

**Comment by werner33** [comment]
@raisedadead Man that is hugely disappointing. I was really hoping to make this a much bigger part of our program. 

Thank you for letting me know. Let me know if I can be helpful in the future. 

Thank you.

**Pull request closed without merging**